### PR TITLE
Implement specialize_fail() (Issue #1970)

### PR DESCRIPTION
**Pull request by steven-johnson**
There’s probably a more elegant way to implement this, but it correctly
represents the shape of the API I think is appropriate.

### CODE REVIEW

File: src/Func.cpp [hunk 1]
```diff
@@ -1124,11 +1124,26 @@ Stage Stage::specialize(Expr condition) {
             return Stage(specializations[i].definition, stage_name, dim_vars, storage_dims);
         }
     }
+
+    // Can't add any more specializations after specialize_fail().
+    for (size_t i = 0; i < specializations.size(); i++) {
+        user_assert(specializations[i].failure_message.empty()) << "Cannot add new specializations after specialiation_fail().";
```
Comment (abadams) [inline src/Func.cpp]: specialization_fail -> specialize_fail

Comment (steven-johnson) [inline src/Func.cpp]: Fixed

File: src/Func.cpp [hunk 2]
```diff
@@ -1124,11 +1124,26 @@ Stage Stage::specialize(Expr condition) {
             return Stage(specializations[i].definition, stage_name, dim_vars, storage_dims);
         }
     }
+
+    // Can't add any more specializations after specialize_fail().
+    for (size_t i = 0; i < specializations.size(); i++) {
+        user_assert(specializations[i].failure_message.empty()) << "Cannot add new specializations after specialize_fail().";
```
Comment (psuriana) [inline src/Func.cpp]: If you can't add new specialization after specialize_fail, isn't that enough to check the last specialize in the list then?

File: src/Func.cpp [hunk 3]
```diff
@@ -1124,11 +1124,26 @@ Stage Stage::specialize(Expr condition) {
...
+    }
     const Specialization &s = definition.add_specialization(condition);
 
     return Stage(s.definition, stage_name, dim_vars, storage_dims);
 }
 
+void Stage::specialize_fail(const std::string &message) {
```
Comment (psuriana) [inline src/Func.cpp]: I think this should assert that 'message' is not empty

File: src/Func.cpp [hunk 4]
```diff
@@ -1124,11 +1124,26 @@ Stage Stage::specialize(Expr condition) {
...
+    const vector<Specialization> &specializations = definition.specializations();
+    for (size_t i = 0; i < specializations.size(); i++) {
+        user_assert(specializations[i].failure_message.empty()) << "Only one specialize_fail() may be defined per Stage.";
```
Comment (psuriana) [inline src/Func.cpp]: Probably enough to just check the last element in the list

File: test/correctness/specialize.cpp [hunk 5]
```diff
@@ -609,6 +609,32 @@ int main(int argc, char **argv) {
         _halide_user_assert(vector_store_lanes == 32);
     }
 
+    {
+        Var x;
+        Param<int> p;
+
+        // Check that if we promote a final const-true specialize, we keep the
```
Comment (psuriana) [inline test/correctness/specialize.cpp]: Is this a stale comment?

Comment (steven-johnson) [inline test/correctness/specialize.cpp]: Fixed.

File: src/Func.cpp (hunk 4 above)
Comment (steven-johnson) [inline src/Func.cpp]: Done

File: src/Func.cpp (hunk 3 above)
Comment (steven-johnson) [inline src/Func.cpp]: Indeed it should, and at one point it did and somehow got lost. Fixed.

File: src/Func.cpp (hunk 2 above)
Comment (steven-johnson) [inline src/Func.cpp]: Done

**Comment by steven-johnson** [comment]
This is now passing. How do we feel about it -- do we want more thinking on the model and/or implementation?

**Comment by abadams** [comment]
The wart to fix is the AssertStmt flag. Default args on IR node make functions are risky - I've been avoiding them. I think any mutation of an AssertStmt is going to drop that flag.

What alternatives are there to a flag? We could only warn on unconditional assert(false)s, but I think a lot of the interesting assertions are inside an if (not bounds inference mode)... Hrm. We could sniff the message to see if it's an unhandled specialize fail call, and not warn. We could have some other way in the IR to return with failure...

Probably sniffing the message to see if it's a call to halide_error_specialize_fail is simplest, but seems like a hack. Removing the default value for the flag and fixing the resulting syntax errors is possibly better. I'd be surprised if there were mutations of AssertStmts outside of Simplify and the base case in IRMutator.cpp

**Comment by steven-johnson** [comment]
> any mutation of an AssertStmt is going to drop that flag

Interesting, I didn't realize that mutation-in-place of Expr/Stmt was something generally done -- I thought the preferred mode was to "mutate" via creating a new Expr/Stmt and replacing the old one. So, yeah, in that case, good point.

> Removing the default value for the flag and fixing the resulting syntax errors is possibly better

Yeah, not hard. (If I took this route I'd use an enum rather than a bool, but yeah.)

> We could only warn on unconditional assert(false)s

Agreed that we might miss cases here. (Though, to be clear: what existing cases can cause a constant-true assert to be generated?)

> What alternatives are there to a flag?

Off the top of my head:
-- Refactor AssertStmt to be based on top of another, unconditional error primitive, and use that new primitive here instead
-- Create a magic value of "constant true" that won't generate the warning. (NO! NO! NO! Just kidding.)

**Comment by abadams** [comment]
There are no mutations-in-place. I meant when a new one is constructed from an existing one, I don't think that flag will be copied over. See IRMutator.cpp line 191. Because all IR nodes are const, this reconstruction happens whenever one of the owned Exprs changes.

**Comment by steven-johnson** [comment]
> Probably sniffing the message to see if it's a call to halide_error_specialize_fail is simplest, but seems like a hack

Both of these things are true; despite the hackiness, I changed it to do the message sniffing, because propagating a new argument into AssertStmt::make() that is used only for such an odd corner case felt like noise. PTAL.

**Comment by abadams** [comment]
LGTM

**Pull request merged**